### PR TITLE
fix bug: incorrect row count for value scan node (#15941)

### PR DESCRIPTION
**Pull request by ouyuanning**
fix bug: incorrect row count for value scan node

Approved by: @badboynt1, @aunjgr

## What type of PR is this?

- [ ] API-change
- [x] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue https://github.com/matrixorigin/MO-Cloud/issues/3214

## What this PR does / why we need it:
fix bug: incorrect row count for value scan node

### CODE REVIEW

**Review by aunjgr — APPROVED** [review]

**Comment by matrix-meow** [comment]
@ouyuanning Thanks for your contributions!

**Here are review comments for file pkg/pb/plan/plan.pb.go:**
### Pull Request Review:

#### Title:
The title of the pull request is clear and concise, indicating that it fixes a bug related to an incorrect row count for a value scan node.

#### Body:
The body of the pull request provides additional context by specifying the type of PR (BUG), the related issue number, and a brief description of the bug fix. It also lists the approvers for the PR.

#### Changes in `pkg/pb/plan/plan.pb.go`:
Since there is no actual diff provided in the pull request, it is challenging to assess the specific changes made to address the bug. It would be beneficial to have the actual code changes visible in the pull request diff to provide a more detailed review.

### Suggestions for Improvement:

1. **Include Code Changes:** It is crucial to include the actual code changes in the pull request diff to allow for a thorough review of the modifications made to address the bug. Without the code changes, it is challenging to provide specific feedback on the implementation.

2. **Detailed Description:** While the title and body provide a general overview of the bug fix, adding more details about the root cause of the issue and the specific changes implemented to resolve it would enhance the clarity of the PR.

3. **Security Considerations:** Ensure that the bug fix does not introduce any security vulnerabilities. Performing a security review of the changes, especially when handling row counts or data scanning, is essential to prevent potential security risks.

4. **Testing:** It is recommended to include relevant test cases to cover the bug fix and ensure that the issue is fully resolved. Automated tests can help in verifying the correctness of the fix and prevent regressions.

5. **Optimization:** If possible, consider optimizing the code changes for better performance or readability. Refactoring the code to improve efficiency or maintainability can be beneficial in the long run.

### Overall, while the pull request addresses a bug related to incorrect row count for a value scan node, it would be beneficial to include the actual code changes, provide more detailed descriptions, consider security implications, add relevant tests, and optimize the code for better quality.

**Here are review comments for file pkg/sql/plan/build_constraint_util.go:**
### Pull Request Review:

#### Title and Body:
The title and body of the pull request are clear and concise, indicating that the PR aims to fix a bug related to an incorrect row count for a value scan node. The PR is approved by relevant team members, and it references the specific issue it addresses.

#### Changes in `pkg/sql/plan/build_constraint_util.go`:
1. **Problem - Incorrect Row Count Update:**
   - The change in the pull request updates the row count for a value scan node by setting `rowsetData.RowCount` to the length of `slt.Rows`.
   - **Issue:** Directly setting `rowsetData.RowCount` to the length of `slt.Rows` may lead to incorrect row count calculations if `slt.Rows` is not the actual number of rows processed by the value scan node.
   - **Suggestion:** Instead of setting `rowsetData.RowCount` directly to `len(slt.Rows)`, consider calculating the actual number of rows processed by the value scan node accurately. This can prevent potential inconsistencies in row count reporting.

2. **Optimization Opportunity:**
   - The code snippet provided in the pull request seems straightforward and does not introduce any performance concerns.
   - **Suggestion:** Since the change is minimal and focused on fixing a bug, there may not be significant optimization opportunities in this specific context.

3. **Security Concerns:**
   - The changes in the pull request do not introduce any apparent security vulnerabilities based on the provided context.

#### General Suggestions:
- Ensure that the row count update accurately reflects the number of rows processed by the value scan node to prevent any discrepancies in reporting.
- Consider adding comments or documentation to explain the rationale behind the row count update for future reference and code maintenance.

### Overall, the pull request addresses a specific bug related to row count calculation for a value scan node. To enhance the code quality, it is recommended to accurately determine the row count processed by the node and provide clear documentation for the changes made.

**Here are review comments for file pkg/sql/plan/deepcopy.go:**
### Pull Request Review:

#### Title and Body:
The title of the pull request is clear and concise, indicating that it fixes a bug related to an incorrect row count for a value scan node. The body provides additional context by mentioning the issue it fixes and the approval from relevant team members.

#### Changes in deepcopy.go:
1. **Problem - Incorrect Row Count Handling:**
   - The change in deepcopy.go modifies the DeepCopyNode function to include the `RowCount` field from the original `RowsetData` when creating a deep copy.
   - The issue arises from the fact that the `RowCount` field is copied directly from the original `node.RowsetData`, which can lead to unexpected behavior if the `RowCount` is modified in the copied node.
  
2. **Security Concern - Data Integrity:**
   - Copying the `RowCount` directly from the original node can potentially lead to data integrity issues if the `RowCount` is modified independently in the copied node. This can result in discrepancies and incorrect processing of data.
  
3. **Optimization Suggestion:**
   - To address the problem and improve the code, it is recommended to ensure that the `RowCount` field is handled appropriately during the deep copy process. Instead of directly copying the `RowCount`, consider recalculating it based on the actual data in the copied node to maintain data consistency.

#### Suggestions for Improvement:
1. **Data Consistency:**
   - When performing a deep copy of the `RowsetData`, calculate the `RowCount` based on the actual data present in the copied node rather than directly copying it. This approach ensures data consistency and avoids potential discrepancies.
  
2. **Code Optimization:**
   - Refactor the deepcopy logic to accurately handle the `RowCount` field during the copy operation. This optimization will enhance the reliability and maintainability of the codebase.

3. **Testing:**
   - It is advisable to include test cases that cover scenarios where the `RowCount` field is crucial to ensure that the changes do not introduce regressions or data integrity issues.

By addressing the data integrity concerns and optimizing the deepcopy logic, the codebase can be improved in terms of reliability and maintainability.

Overall, the pull request addresses a specific bug, but the handling of the `RowCount` field requires attention to prevent potential issues related to data integrity.

**Here are review comments for file pkg/sql/plan/stats.go:**
### Pull Request Review: fix bug: incorrect row count for value scan node (#15941)

#### Problem 1: Inconsistent Variable Naming
- In the `stats.go` file, the variable `rowCount` is being used to represent the number of rows, but the calculation and assignment of this variable are not consistent.
- The calculation of `rowCount` is based on `node.RowsetData.RowCount`, which is correct, but the variable name `rowCount` itself is misleading as it implies a count of rows, not the total count of values.
- **Suggestion:** Rename the variable `rowCount` to something more descriptive like `totalValueCount` to accurately reflect its purpose and improve code readability.

#### Problem 2: Redundant Code
- The original code snippet shows that the variable `colsData` is being retrieved from `node.RowsetData.Cols`, but it is not used in the subsequent code.
- The `len(colsData[0].Data)` calculation is redundant and unnecessary for determining the row count.
- **Suggestion:** Remove the unused `colsData` variable and the redundant calculation to simplify the code and improve its clarity.

#### Problem 3: Magic Number Usage
- The calculation `int32(rowCount/float64(options.DefaultBlockMaxRows) + 1)` contains a magic number `options.DefaultBlockMaxRows` which lacks context and may lead to maintenance issues in the future.
- Magic numbers make the code less maintainable and harder to understand.
- **Suggestion:** Define a named constant for `options.DefaultBlockMaxRows` to provide clarity and maintainability. Replace the magic number with this constant in the calculation.

#### Problem 4: Lack of Comments
- The code lacks comments explaining the purpose of the fix and the logic behind the calculations.
- Comments are essential for understanding the code's intent, especially for complex calculations or bug fixes.
- **Suggestion:** Add comments to explain the purpose of the fix, the calculation of `rowCount`, and any significant logic in the code to enhance readability and maintainability.

#### Security Concern:
- No security concerns were identified in the provided code changes.

#### Overall Optimization Suggestions:
- Improve variable naming for clarity and consistency.
- Remove redundant code to simplify the logic.
- Replace magic numbers with named constants for better maintainability.
- Add comments to explain the purpose and logic of the code changes.

By addressing the identified issues and implementing the suggested improvements, the code quality, readability, and maintainability of the `stats.go` file can be enhanced.

**Here are review comments for file proto/plan.proto:**
### Pull Request Review:

#### Title and Body:
The title and body of the pull request are clear and concise, indicating that the purpose of the PR is to fix a bug related to an incorrect row count for a value scan node. The PR is approved by the relevant team members, and it references the specific issue it addresses.

#### Changes in `proto/plan.proto`:
1. **Addition of `row_count` field in `RowsetData` message:**
   - The addition of the `int32 row_count = 2;` field in the `RowsetData` message seems to address the bug related to the incorrect row count for a value scan node.
   - **Issue:** The addition of the `row_count` field alone may not be sufficient to fix the bug. It is essential to ensure that this field is correctly populated and updated to reflect the actual row count.
   - **Suggestion:** Verify that the `row_count` field is properly calculated and updated whenever rows are added or removed in the `RowsetData` message. Ensure that the logic for updating this field is accurate to prevent future discrepancies.

#### Suggestions for Optimization:
1. **Documentation:**
   - Consider adding comments or documentation within the `proto/plan.proto` file to explain the purpose and usage of the `row_count` field. This will help developers understand its significance and how it should be maintained.

2. **Testing:**
   - It is crucial to include tests that cover scenarios where the `row_count` field is updated to ensure its correctness under various conditions. Adding unit tests specifically targeting the functionality related to the row count can help prevent regressions.

3. **Code Review:**
   - Conduct a thorough code review to ensure that the changes made in this PR do not introduce any unintended side effects or break existing functionality. Pay close attention to how the `row_count` field is utilized throughout the codebase.

4. **Validation:**
   - Before merging the PR, validate the fix by testing it in a real-world scenario to confirm that the row count is now accurately reflected for the value scan node.

#### Security Implications:
Based on the provided diff, there are no apparent security implications. However, it is always recommended to follow secure coding practices and conduct security reviews when making changes to the codebase.

By addressing the suggestions mentioned above and ensuring the correctness and completeness of the fix, the codebase can be improved, and the bug related to the incorrect row count for the value scan node can be effectively resolved.

**Review by badboynt1 — APPROVED** [review]

**Review by sukki37 — APPROVED** [review]

**Pull request merged**
